WJ-685 & WJ-66 (Partially) — Repurposing Site Tags as a Tag Restriction System

Yossipossi1 opened the pull request:
This pull request is designed to repurpose the deprecated "Site Tags" feature for an ability to restrict tags to a specific list, as detailed in [WJ-66](https://scuttle.atlassian.net/browse/WJ-66) and [WJ-685](https://scuttle.atlassian.net/browse/WJ-685). This option is toggle-able in the general site settings. Most of the code is in Ozone, though some of the it is in Laravel. 

Obviously this is not the end of working on tags, but it's a good first step to get the framework in place for future development. I plan on revisiting this in the future to convert all code involving tags to Laravel.

Yossipossi1 committed:
repurpose site tags to validate page tags

Site Tags do not exist in the modern version of Wikidot. I have repurposed them to be a list of possible tags for a wiki which every page validates against when adding new tags.

Yossipossi1 committed:
repurpose site tags to validate page tags

Site Tags do not exist in the modern version of Wikidot. I have repurposed them to be a list of possible tags for a wiki which every page validates against when adding new tags.

Yossipossi1 committed:
Update PageTagsModule.php

Yossipossi1 committed:
Changed SiteTags to AllowedTags and Removed Tag Size Limit

danieltharp committed:
Add default to allowed tags bool

Yossipossi1 committed:
The "AllowedTags" feature is now funtioning, allowing for sites to enable/disable them, and checking tags against them to ensure they don't conflict. Further work needs to be done with tags, but this works well for now.

Yossipossi1 committed:
Minor Fix

Accidentally removed something I shouldn't have.

Yossipossi1 committed:
Removing Duplicate Code

For some reason I had a bunch of duplicate code, likely left over from debugging.

Yossipossi1 committed:
Removing Site Tags Notice From Manage Site

There was a notice explaining Site Tags at the bottom of the page; as Site Tags are now Allowed Tags, I've removed this notice.

Yossipossi1 committed:
...Missed a part.

emmiegit commented on web/php/Actions/ManageSiteAction.php:
```diff
@@ -438,8 +438,9 @@ public function saveGeneralEvent($runData)
-
         $description = trim($pl->getParameterValue("description"));
+        $enableAllowedTags = trim($pl->getParameterValue("enable_allowed_tags"));
+        $enableAllowedTags = !empty($enableAllowedTags) ? 'true' : 'false';
```
Why are these strings instead of booleans?

emmiegit commented on web/php/DB/AllowedTags.php:
```diff
@@ -0,0 +1,32 @@
+        foreach ($dbTags as $dbTag) {
+            $allowedTags .= htmlspecialchars($dbTag->getTag()).' ';
+        }
+        return trim($allowedTags);
```
Why is this using space-separated tags in a string instead of `array<string>`? In the uses I can see you immediately `explode()`, so I don't think it makes sense to store it this way.

emmiegit commented on web/php/Actions/WikiPageAction.php:
```diff
@@ -1407,8 +1414,17 @@ public function saveTagsEvent($runData)
+       if($enableAllowedTags == true) {
+            foreach ($pageTagsArray as $tag) {
+                if(!in_array($tag, $allowedTagsList)) {
+                    throw new ProcessException(_('Invalid tags.'), "form_error");
```
I would give it a more specific error message, like "Tag XYZ is not valid for this site."

emmiegit committed:
Revert file modes for empty gitignore files.

emmiegit committed:
Use booleans instead of strings.

Yossipossi1 committed:
Changing Error Message

Changing the error message of Allowed Tags to better reflect the reason for tags being invalid.

Yossipossi1 committed:
1. Reverts the string -> boolean change for the time being, as it doesn't seem to work properly as a boolean (will revisit later; functions regardless for the time being)
2. All AllowedTags queries are now in Laravel.
3. Adds conditional to WikiPageAction.php to only perform AllowedTags check if the tags field isn't empty.

Yossipossi1 committed:
Fixing error message (again).

emmiegit committed:
Remove executable permissions from .gitignore file.

emmiegit commented on web/php/Actions/ManageSiteAction.php:
```diff
@@ -438,8 +438,9 @@ public function saveGeneralEvent($runData)
-
         $description = trim($pl->getParameterValue("description"));
+        $enableAllowedTags = $pl->getParameterValue("enable_allowed_tags");
+        $enableAllowedTags = !empty($enableAllowedTags) ? 'true' : 'false';
```
Since this is intentional (for now), I would leave a comment explaining why these are strings.

emmiegit commented on web/php/Actions/WikiPageAction.php:
```diff
@@ -1407,8 +1414,17 @@ public function saveTagsEvent($runData)
+
+        $allowedTagsList = AllowedTags::getAllowedTags($siteId);
+
+       if($enableAllowedTags == true && !empty($tags)) {
```
Either
```suggestion
       if($enableAllowedTags && !empty($tags)) {
```
or
```suggestion
       if($enableAllowedTags === true && !empty($tags)) {
```

Yossipossi1 committed:
Adding comment + changing conditional slightly
